CIP-????/README.md

```diff
@@ -0,0 +1,156 @@
+  deriving (Pretty) via (PrettyShow ScriptInfo)
+```
+
+```hs
+data ScriptPurpose
+  = MintingScript
+      -- hash of the script being executed (same as polciy)
+      V2.CurrencySymbol
+      -- | 0-based index of the given `CurrencySymbol` in `txInfoMint`
+      Haskell.Integer
+  | SpendingScript
+      -- hash of the script being executed
+      V2.ScriptHash
+      -- | 0-based index of the input being spent in `txInfoInputs`
+      Haskell.Integer
+      V3.TxOutRef
+  | RewardingScript
+      -- V2.Credential -- removed, since implictily it must be a script credential
+      -- hash of the script being executed
+      V2.ScriptHash
+      -- | 0-based index of the withdrawal being verified in `txInfoWdrl`
+      Haskell.Integer
+  | CertifyingScript
+      -- hash of the script being executed
+      V2.ScriptHash
+      -- | 0-based index of the given `TxCert` in `txInfoTxCerts`
+      Haskell.Integer
+      TxCert
+  | VotingScript
+      -- hash of the script being executed
+      V2.ScriptHash
+      -- | 0-based index of the given vote in `txInfoVotes`
+      Haskell.Integer
+      Voter
+  | ProposingScript
+      -- hash of the script being executed
+      V2.ScriptHash
+      -- | 0-based index of the given `ProposalProcedure` in `txInfoProposalProcedures`
+      Haskell.Integer
+      ProposalProcedure
+  deriving stock (Generic, Haskell.Show, Haskell.Eq)
+  deriving anyclass (HasBlueprintDefinition)
+  deriving (Pretty) via (PrettyShow ScriptInfo)
+```
+
+## Rationale: how does this CIP achieve its goals?
+<!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It should describe alternate designs considered and related work. The rationale should provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.
+
+It must also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section should explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.
+-->
+
+This modification would be reflected in the underlying data representation as always having the script hash of the given script as the first field of the purpose, and the index relevant to the purpose as second element.
+
+This allows to optimize for the very common operation of deriving the hash of the script itself onchain, greatly reducing onchain complexity and cost of execution.
+
+The change also propses to add the resovled input in the case of the very common spending purpose, where the majority of the relevant informations is present.
+
+## Path to Active
+
+Changes are reflected on the plutus-ledger-api
+
+### Acceptance Criteria
+<!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->
+
+The change is included in an hardfork
```

Comment:
```suggestion
The change is included in a hardfork.
```
This could use some elaboration but will wait until the Ledger responses are all in & Plutus reps have reviewed with respect to when this might be included (cc @zliu41).